AsyncDecodingSource.h

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) 2020 Open Mobile Platform LLC.
+ * Copyright 2016, The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef ASYNC_DECODING_SOURCE_H_
+#define ASYNC_DECODING_SOURCE_H_
+
+#include <media/stagefright/foundation/AString.h>
+#include <media/stagefright/foundation/Mutexed.h>
+#include <media/stagefright/foundation/AHandlerReflector.h>
+#if ANDROID_MAJOR >= 9
+#include <media/MediaSource.h>
+#else
+#include <media/stagefright/MediaSource.h>
+#endif
+#include <utils/Condition.h>
+#include <utils/StrongPointer.h>
+
+struct ANativeWindow;
+
+namespace android {
+
+struct ALooper;
+struct AMessage;
+class MediaBuffer;
+class MediaBufferBase;
+struct MediaCodec;
+class MetaData;
+class Surface;
+
+class AsyncDecodingSource : public MediaSource {
+public:
+    static sp<AsyncDecodingSource> Create(const sp<MediaSource> &source,
+                    uint32_t flags, const sp<ANativeWindow> &nativeWindow,
+                    const sp<ALooper> &looper, const char *desiredCodec = NULL);
+
+    bool configure(const sp<AMessage> format,
+                   const sp<Surface> surface,
+                   uint32_t flags = 0);
+
+    virtual ~AsyncDecodingSource();
+
+    // starts this source (and its underlying source). |params| is ignored.
+    virtual status_t start(MetaData *params = NULL);
+
+    // stops this source (and its underlying source).
+    virtual status_t stop();
+
+    // returns the output format of this source.
+    virtual sp<MetaData> getFormat();
+
+    // reads from the source. This call always blocks.
+    virtual status_t read(
+#if ANDROID_MAJOR >= 9
+                    android::MediaBufferBase **buffer,
+#else
+                    android::MediaBuffer **buffer,
+#endif
```

Comment:
Maybe add an type alias instead of repeating the preprocessor everywhere this turns up.

Reply:
Done for this and the MediaCodecBuffer/ABuffer too. Much cleaner.